### PR TITLE
[feature] add expressions at the symbollist level

### PR DESCRIPTION
**Pull request by vmora**
Size and Rotation can be defined by an expression for all symbols
composing a marker Width can be defined by an expression for all symbols
composing a line

For marquers, a legend is generated for varying sizes. This allows
multivariate analysis legend in the case of classified/graduated colors.

The line symbol layer as been refactored to avoid code duplication and
expose the offset and offset units in the base class. This refacto was
necessary for this feature to avoid unnecessary dynamic cast. Note that
the added functions in the base class where already defined in all child
classes.

The offset is now set along with size to maintain the relative position
of symbols composing a marker.

An asistant, with preview, is accessible through the data defined button
to help the user define the size expression. Three methods are
available: Frannery, Area and Radius.

Added a widged for use in categorized/classified symbology gui to set
the expression if needed. The assistant is also available from it.

Note: breaking change in the symbolslist widget. The widget ctor now
takes the layer as a parameter, this is necessary for the assistant and
for the data-defined buttons. The widget is used only in one place and
although it's a change to the public interface, the likelyhood of
breaking plugins dependent on this widget seems low enought.

### CODE REVIEW

**Comment by vmora** [comment]
replaces #1961

**Comment by vmora** [comment]
fixing...

File: src/core/layertree/qgslayertreemodellegendnode.cpp [hunk 1]
```diff
@@ -187,11 +188,67 @@ QVariant QgsSymbolV2LegendNode::data( int role ) const
...
+          QImage img = pix.toImage();
+          // find the non-transparent area bbox
+          int xmin = img.width(), xmax = 0, ymin = img.height(), ymax = 0;
+          for ( int x = 0; x < img.width(); x++ )
+          {
+            for ( int y = 0; y < img.height(); y++ )
+            {
+              if ( qAlpha( img.pixel( x, y ) ) )
+              {
+                xmin = x < xmin ? x : xmin;
```
Comment (nyalldawson) [inline src/core/layertree/qgslayertreemodellegendnode.cpp]: May as well use qMin/qMax here

File: src/core/layertree/qgslayertreemodellegendnode.cpp [hunk 2]
```diff
@@ -187,11 +188,67 @@ QVariant QgsSymbolV2LegendNode::data( int role ) const
         context.setRendererScale( scale );
         context.setMapToPixel( QgsMapToPixel( mupp ) ); // hope it's ok to leave out other params
 
-        pix = QgsSymbolLayerV2Utils::symbolPreviewPixmap( mItem.symbol(), iconSize, validData ? &context : 0 );
+        // crop
+        if ( mItem.symbol()->type() == QgsSymbolV2::Marker && mCrop )
+        {
+          pix = QgsSymbolLayerV2Utils::symbolPreviewPixmap( mItem.symbol(), QSize( 512, 512 ), validData ? &context : 0 );
```
Comment (nyalldawson) [inline src/core/layertree/qgslayertreemodellegendnode.cpp]: this section should be moved out to its own static function in QgsSymbolLayerV2Utils, it's a useful routine to have widely available

File: src/core/layertree/qgslayertreemodellegendnode.cpp [hunk 3]
```diff
@@ -187,11 +188,67 @@ QVariant QgsSymbolV2LegendNode::data( int role ) const
...
+          }
+          pix = QPixmap::fromImage( img.copy( xmin, ymin, xmax - xmin, ymax - ymin ) );
+        }
+        else if ( mItem.symbol()->type() == QgsSymbolV2::Line && mCrop )
+        {
+          pix = QgsSymbolLayerV2Utils::symbolPreviewPixmap( mItem.symbol(), QSize( mIconSize.width(), 512 ), validData ? &context : 0 );
+          QImage img = pix.toImage();
+          // find the non-transparent area bbox
+          int ymin = img.height(), ymax = 0;
+          for ( int y = 0; y < img.height(); y++ )
```
Comment (nyalldawson) [inline src/core/layertree/qgslayertreemodellegendnode.cpp]: this loop makes the assumption that a line symbol is constant along it's length. That's not necessarily the case, eg for a marker line symbol.

File: src/core/layertree/qgslayertreemodellegendnode.h [hunk 4]
```diff
@@ -162,6 +162,11 @@ class CORE_EXPORT QgsSymbolV2LegendNode : public QgsLayerTreeModelLegendNode
 
     virtual void invalidateMapBasedData() override;
 
+    void setIconSize( const QSize& sz ) { mIconSize = sz; }
```
Comment (nyalldawson) [inline src/core/layertree/qgslayertreemodellegendnode.h]: these methods need documentation

File: src/core/layertree/qgslayertreemodellegendnode.h [hunk 5]
```diff
@@ -162,6 +162,11 @@ class CORE_EXPORT QgsSymbolV2LegendNode : public QgsLayerTreeModelLegendNode
...
+    QSize iconSize() const { return mIconSize; }
+
+    void setCrop( bool isCropRequired ) { mCrop = isCropRequired; }
```
Comment (nyalldawson) [inline src/core/layertree/qgslayertreemodellegendnode.h]: I'd suggest adding a getter for mCrop too. There's nothing more frustrating then a missing getter when your project depends on it ;)

File: src/core/qgsdatadefined.cpp [hunk 6]
```diff
@@ -43,6 +43,18 @@ QgsDataDefined::QgsDataDefined( const QgsExpression * expression )
   mExpressionPrepared = false;
 }
 
+QgsDataDefined::QgsDataDefined( const QString & expressionString )
+{
+  QgsExpression expression( expressionString );
+  mActive = expression.rootNode();
+  mUseExpression = expression.rootNode() && !dynamic_cast<const QgsExpression::NodeColumnRef*>( expression.rootNode() );
```
Comment (nyalldawson) [inline src/core/qgsdatadefined.cpp]: I'd suggest moving this check to something like QgsExpression::isSingleField() , and adding a few unit tests for it.

File: src/core/qgsscaleexpression.cpp [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  return "";
+}
+
+
+double QgsScaleExpression::size( double value ) const
+{
+  switch ( mType )
+  {
+    case Linear: return mMinSize + ( qBound( mMinValue, value, mMaxValue ) - mMinValue ) * ( mMaxSize - mMinSize ) / ( mMaxValue - mMinValue );
+    case Area: return pow( qBound( mMinValue, value, mMaxValue ) - mMinValue, .5 ) * ( mMaxSize - mMinSize ) / pow( mMaxValue - mMinValue, .5 );
```
Comment (nyalldawson) [inline src/core/qgsscaleexpression.cpp]: should be qPow instead of pow

File: src/core/symbology-ng/qgssymbollayerv2utils.h [hunk 8]
```diff
@@ -426,6 +426,13 @@ class CORE_EXPORT QgsSymbolLayerV2Utils
      */
     static QString fieldOrExpressionFromExpression( QgsExpression* expression );
 
+    /** Computes a sequence of about 'classes' equally spaced round values
+     *  which cover the range of values from 'minimum' to 'maximum'.
+     *  The values are chosen so that they are 1, 2 or 5 times a power of 10.
+     * @note added in 2.10
+     */
+    static QList<double> prettyBreaks( double minimum, double maximum, int classes );
```
Comment (nyalldawson) [inline src/core/symbology-ng/qgssymbollayerv2utils.h]: needs to be added to sip bindings too

**Comment by nyalldawson** [comment]
@vmora Just a note, I'm only part way through this. It's a big PR :)

File: src/gui/qgsdatadefinedbutton.h [hunk 9]
```diff
@@ -167,6 +181,12 @@ class GUI_EXPORT QgsDataDefinedButton: public QToolButton
     void clearCheckedWidgets() { mCheckedWidgets.clear(); }
 
     /**
+     * Set an assistant to define expression
+     * takes ownership
+     */
+    void setAssistant( QgsDataDefinedAssistant * assistant ) { mAssistant.reset( assistant ); }
```
Comment (nyalldawson) [inline src/gui/qgsdatadefinedbutton.h]: probably the parent of mAssistant should be set to "this" here too. That will help ensure that the dialog gets displayed in the right position (eg for multi monitor setups)

File: src/gui/qgsdatadefinedbutton.h [hunk 10]
```diff
@@ -15,16 +15,30 @@
 #ifndef QGSDATADEFINEDBUTTON_H
 #define QGSDATADEFINEDBUTTON_H
 
-#include "qgsfield.h"
+#include <qgsfield.h>
+#include <qgsdatadefined.h>
```
Comment (nyalldawson) [inline src/gui/qgsdatadefinedbutton.h]: Is this needed in the header?

File: src/gui/qgsdatadefinedbutton.cpp [hunk 11]
```diff
@@ -394,6 +404,21 @@ void QgsDataDefinedButton::showDescriptionDialog()
   mv->exec();
 }
 
+void QgsDataDefinedButton::showAsssistant()
+{
+  if ( mAssistant->exec() == QDialog::Accepted )
+  {
+    const QString newExp = mAssistant->expressionText();
```
Comment (nyalldawson) [inline src/gui/qgsdatadefinedbutton.cpp]: Why not make QgsDataDefinedAssistant return a QgsDataDefined instead of just a string? That way assistants could be created which fields, deactivated expressions, etc. It would also be more future proof when QgsDataDefined is extended to support additional features (such as time based/animation support).

File: src/gui/qgsdatadefinedbutton.h [hunk 12]
```diff
@@ -15,16 +15,30 @@
...
 #include <QMap>
 #include <QPointer>
 #include <QToolButton>
+#include <QScopedPointer>
 
 class QgsVectorLayer;
 class QgsDataDefined;
 
 /** \ingroup gui
+ * \class Qgs
```
Comment (nyalldawson) [inline src/gui/qgsdatadefinedbutton.h]: All these added methods/classes need to be added to the SIP bindings too, and should have "added in ..." notes.

**Comment by nyalldawson** [comment]
@vmora I think we could also speed up merging this if we commit it piece-by-piece. Eg, I **really** like the "assistant" changes you've made to QgsDataDefinedButton, so if the small fixes noted were made to that then we could merge that portion right away.

File: src/core/symbology-ng/qgssymbollayerv2utils.cpp [hunk 13]
```diff
@@ -3781,3 +3781,146 @@ QString QgsSymbolLayerV2Utils::fieldOrExpressionFromExpression( QgsExpression* e
   return expression->expression();
 }
 
+QList<double> QgsSymbolLayerV2Utils::prettyBreaks( double minimum, double maximum, int classes )
```
Comment (nyalldawson) [inline src/core/symbology-ng/qgssymbollayerv2utils.cpp]: this change (moving _prettyBreaks) to QgsSymbolLayerV2Utils could be split into it's own commit - that way we could push these changes straight away and help decrease the size of this PR

File: src/core/layertree/qgslayertreemodellegendnode.cpp (hunk 1 above)
Comment (vmora) [inline src/core/layertree/qgslayertreemodellegendnode.cpp]: done

File: src/core/layertree/qgslayertreemodellegendnode.cpp (hunk 3 above)
Comment (vmora) [inline src/core/layertree/qgslayertreemodellegendnode.cpp]: fixed with a loop on x too.

File: src/core/layertree/qgslayertreemodellegendnode.h (hunk 4 above)
Comment (vmora) [inline src/core/layertree/qgslayertreemodellegendnode.h]: done

File: src/core/layertree/qgslayertreemodellegendnode.h (hunk 5 above)
Comment (vmora) [inline src/core/layertree/qgslayertreemodellegendnode.h]: done

File: src/core/qgsscaleexpression.cpp (hunk 7 above)
Comment (vmora) [inline src/core/qgsscaleexpression.cpp]: done

File: src/core/symbology-ng/qgssymbollayerv2utils.h (hunk 8 above)
Comment (vmora) [inline src/core/symbology-ng/qgssymbollayerv2utils.h]: done

File: src/gui/qgsdatadefinedbutton.h (hunk 9 above)
Comment (vmora) [inline src/gui/qgsdatadefinedbutton.h]: done

File: src/gui/qgsdatadefinedbutton.h (hunk 10 above)
Comment (vmora) [inline src/gui/qgsdatadefinedbutton.h]: nope, you're right. done

File: src/gui/qgsdatadefinedbutton.cpp (hunk 11 above)
Comment (vmora) [inline src/gui/qgsdatadefinedbutton.cpp]: done.

File: src/core/layertree/qgslayertreemodellegendnode.cpp (hunk 2 above)
Comment (vmora) [inline src/core/layertree/qgslayertreemodellegendnode.cpp]: done, function crop in QgsSymbolLayerV2Utils

File: src/core/layertree/qgslayertreemodellegendnode.cpp (hunk 3 above)
Comment (vmora) [inline src/core/layertree/qgslayertreemodellegendnode.cpp]: now uses function crop, wich is the same as the fix and avoids duplication.

**Comment by nyalldawson** [comment]
@vmora I've just committed some portions of this directly, to help reduce the size of this PR. There's a lot  happening here and I'm having trouble getting my head around some of it. Hope that's OK! Can you rebase this branch against master and I'll review the remaining changes?

Thanks!

**Comment by vmora** [comment]
@nyalldawson Done.

**Comment by vmora** [comment]
@nyalldawson oups, not yet

**Comment by vmora** [comment]
@nyalldawson now done, really :)

**Comment by nyalldawson** [comment]
@vmora I've just pushed some more components of this, thanks! I made a few changes to QgsScaleExpression:
- ran the code through scripts/prepare-commit.sh to fix spacing
- fixed the copyright header
- added python bindings
- also fixed some of the maths in QgsScaleExpression::size - some of the types weren't adding mMinSize to the returned value
- some other QString related optimisations

Hope that's ok! Don't worry about rebasing after this, there's not too much left to sort through

File: src/core/symbology-ng/qgssymbolv2.h [hunk 14]
```diff
@@ -280,10 +281,26 @@ class CORE_EXPORT QgsMarkerSymbolV2 : public QgsSymbolV2
...
+     */
+    QString angleExpression() const;
 
     void setSize( double size );
-    double size();
+    double size() const;
+
+    /** Set "en masse" expression for size
+     */
+    void setSizeExpression( const QString & exprStr );
```
Comment (nyalldawson) [inline src/core/symbology-ng/qgssymbolv2.h]: these methods (setSizeExpression, sizeExpression, setWidthExpression, widthExpression) should accept/return a QgsDataDefined rather than QStrings. This will avoid the current situation where deactivating the size/width data defined button results in lose of the expression, and will future proof this work for when QgsDataDefined is enhanced in future (eg, there's plans for QgsDataDefined to allow for animation support, etc).

File: src/core/symbology-ng/qgssymbolv2.cpp [hunk 15]
```diff
@@ -35,6 +37,33 @@
 
 #include <cmath>
 
+inline
+QString rotateEnMasse( double additionalRotation, const QString & exprStr )
```
Comment (nyalldawson) [inline src/core/symbology-ng/qgssymbolv2.cpp]: as per an earlier version of this PR, can these "enMasse" methods/classes all be renamed to avoid the use of "enMasse"? Something like "rotateWholeSymbol" or "rotateAllSymbolLayers" would be easier to understand and result in a nicer API

File: src/gui/symbology-ng/qgssymbolslistwidget.h [hunk 16]
```diff
@@ -30,7 +30,7 @@ class GUI_EXPORT QgsSymbolsListWidget : public QWidget, private Ui::SymbolsListW
     Q_OBJECT
 
   public:
-    QgsSymbolsListWidget( QgsSymbolV2* symbol, QgsStyleV2* style, QMenu* menu, QWidget* parent );
+    QgsSymbolsListWidget( const QgsVectorLayer * layer, QgsSymbolV2* symbol, QgsStyleV2* style, QMenu* menu, QWidget* parent );
```
Comment (nyalldawson) [inline src/gui/symbology-ng/qgssymbolslistwidget.h]: unfortunately this API break isn't allowed. You'll need to make the layer optional and add checks to qgssymbolslistwidget.cpp so that all references to mLayer are optional. Looks like this only affects the data defined buttons, so those could be either disable or hidden if no layer is specified...

File: src/gui/symbology-ng/qgssymbolv2selectordialog.cpp [hunk 17]
```diff
@@ -110,31 +110,31 @@ class SymbolLayerItem : public QStandardItem
     QVariant data( int role ) const override
     {
       if ( role == Qt::DisplayRole || role == Qt::EditRole )
+  {
```
Comment (nyalldawson) [inline src/gui/symbology-ng/qgssymbolv2selectordialog.cpp]: remember to run scripts/prepare_commit.sh to fix the spacing before committing, that should avoid unnecessary changes like these

File: src/core/layertree/qgslayertreemodellegendnode.h [hunk 18]
```diff
@@ -162,6 +162,18 @@ class CORE_EXPORT QgsSymbolV2LegendNode : public QgsLayerTreeModelLegendNode
 
     virtual void invalidateMapBasedData() override;
 
+    //! Set the icon size (icon may be smaller if croping is enabled)
```
Comment (nyalldawson) [inline src/core/layertree/qgslayertreemodellegendnode.h]: @wonder-sk do you mind looking at these legend related changes? ( QgsMapLayerLegend + QgsLayerTreeModelLegendNode)

File: src/gui/symbology-ng/qgssymbolslistwidget.h (hunk 16 above)
Comment (vmora) [inline src/gui/symbology-ng/qgssymbolslistwidget.h]: done

File: src/gui/symbology-ng/qgssymbolv2selectordialog.cpp (hunk 17 above)
Comment (vmora) [inline src/gui/symbology-ng/qgssymbolv2selectordialog.cpp]: I usually do, sorry for that. Fixed.

File: src/core/symbology-ng/qgssymbolv2.cpp (hunk 15 above)
Comment (vmora) [inline src/core/symbology-ng/qgssymbolv2.cpp]: done

File: src/core/symbology-ng/qgssymbolv2.h (hunk 14 above)
Comment (vmora) [inline src/core/symbology-ng/qgssymbolv2.h]: Hi @nyalldawson, 

I don't mind changing the API, but the underlying storage is expressions, so it won't fix the problem right away. Or do you suggest I change mDataDefinedProperties to be a map of QgsDataDefined ? Because that's a huge change.

As a side note, I think that some parts of QgsDataDefined are necessary (like activate/deactivate) but some parts should be handled by expression, like deciding if it's a simple field or imposing the return type (improvement of qgsexpression required).

File: src/core/qgsmaplayerlegend.cpp [hunk 19]
```diff
@@ -198,11 +198,33 @@ QList<QgsLayerTreeModelLegendNode*> QgsDefaultVectorLayerLegend::createLayerTree
...
+      n->setCrop( false );
+      n->setIconSize( pix.size() );
+    }
+  }
+
+  foreach ( QgsSymbolV2LegendNode* n, symbolNodes )
+  {
+    Q_ASSERT( widthMax > 0 );
+    n->setIconSize( QSize( widthMax + 2, n->iconSize().rheight() + 2 ) );
+    n->setCrop( false );
```
Comment (wonder-sk) [inline src/core/qgsmaplayerlegend.cpp]: If I understand the code correctly, it temporarily turns on cropping to get the minimum size of each symbol, calculates the max width and finally sets the icon size for all icons to match the width. After initialization, cropping is always disabled.

- could you remove crop()/setCrop() from the API and just have a call minimumIconSize() to get that cropped icon size? I would prefer to keep such code away from data() implementation as it is unclear the cropping is enabled just temporarily.
- why +2 for icon sizes? (better always comment why magic numbers are used)
- how about rule-based renderer? it shows legend in a tree, so I guess there should be different max width for each sub-tree (otherwise one wide icon somewhere in the tree will make everything wide)

File: src/gui/symbology-ng/qgssymbolslistwidget.h [hunk 20]
```diff
@@ -30,7 +30,7 @@ class GUI_EXPORT QgsSymbolsListWidget : public QWidget, private Ui::SymbolsListW
     Q_OBJECT
 
   public:
-    QgsSymbolsListWidget( QgsSymbolV2* symbol, QgsStyleV2* style, QMenu* menu, QWidget* parent );
+    QgsSymbolsListWidget( QgsSymbolV2* symbol, QgsStyleV2* style, QMenu* menu, QWidget* parent, const QgsVectorLayer * layer );
```
Comment (nyalldawson) [inline src/gui/symbology-ng/qgssymbolslistwidget.h]: layer will need a default value of 0 to avoid the API break

File: src/core/symbology-ng/qgssymbolv2.h (hunk 14 above)
Comment (nyalldawson) [inline src/core/symbology-ng/qgssymbolv2.h]: @vmora for new methods like the en masse size these should definitely be stored as QgsDataDefined. There's two main reasons for this - one is that the behaviour of QgsDataDefinedButtons should be consistent between symbology and labeling/composer, and this requires storing the extra information like active/deactive which requires QgsDataDefined. The more important reason though is that it future proofs this work. I'm currently working on enhancements to QgsDataDefined to allow a system of inheritable properties which are evaluated against expression contexts (you can see some early discussion here https://github.com/qgis/QGIS-Enhancement-Proposals/pull/12 and an early version of the expression context work here https://github.com/nyalldawson/QGIS/tree/exp_context8 and here https://github.com/nyalldawson/QGIS/commit/cb5f244fc92c60418a25f8fa45fae4b95fb717c8 ). This work will make implementing things like this PR much easier, but unfortunately it's still a while off. In the meantime all new methods should utilise QgsDataDefined so they'll be ready to flip across.

I'm more than happy to help with this -- if it's not done now it will cause me a lot of work in future ;)

(btw, I started work on porting symbology to use QgsDataDefined some time ago, but haven't had a chance to finish this - the work in progress is here https://github.com/nyalldawson/QGIS/commit/6126e09418b49ec8a2258a43a85849fb80148c0b ).

Comment (nyalldawson) [inline src/core/symbology-ng/qgssymbolv2.h]: FYI - I'm working on fixing up that switch to QgsDataDefined for symbology. Should be committed sometime in the next 48 hours, so maybe wait for that to land as it should simplify this...

Comment (nyalldawson) [inline src/core/symbology-ng/qgssymbolv2.h]: @vmora ok, I've pushed that work. Let me know if you need a hand porting this code over to QgsDataDefined

File: src/gui/symbology-ng/qgssymbolslistwidget.h (hunk 20 above)
Comment (vmora) [inline src/gui/symbology-ng/qgssymbolslistwidget.h]: Yes, off course, sorry for this mistake.

File: src/core/qgsmaplayerlegend.cpp (hunk 19 above)
Comment (vmora) [inline src/core/qgsmaplayerlegend.cpp]: @wonder-sk, you understand the code correctly, I removed rop()/setCrop() and replaced by minimumIconSize() and the code is a lot nicer like that (876603aa2).

I don't know how to handle the tree representation, could you give me some advice/code example please ?

File: src/core/symbology-ng/qgssymbolv2.h (hunk 14 above)
Comment (vmora) [inline src/core/symbology-ng/qgssymbolv2.h]: @nyalldawson I fixed the missing default value and made the changes recommender by @wonder-sk today.

I'm going to rebase the code against master and use QgsDataDefined.

**Comment by wonder-sk** [comment]
@vmora for the legend tree representation... the legend nodes may have a parent legend node assigned (actually key (ID) of the parent rule) - that information is used to build the tree from the flat list of nodes. So one basically needs to group legend nodes with the same parent and do the calculation of icon size in those groups.

Hope that helps... the key is to use QgsLayerTreeModelLegendNode::data(ParentRuleKeyRole) to get the parents. So far only the rule-based renderer supports that.

**Comment by vmora** [comment]
@wonder-sk thanks, that helps a lot, 44ecead6abf takes care of that.

@nyalldawson the previous commit (changing the crop to minimumIconSize()) broke the size assistant. I'm fixing that.

**Comment by vmora** [comment]
@nyalldawson I fixed the assistant, but I don't see the legend for varying size anymore. I think it's related to the rebase against master with your change for QgsDatadefined.

**Comment by vmora** [comment]
I found the problem: the mExpression creation is deferred now, so that the QgsSymbolLayerV2::dataDefinedProperty now return NULL even if there is a mExpressionString.

I'm changing the interface (and implementation to avoid the use of the deprecated symbollayer interface) and it'll fix the problem.

But I think that storing the string AND the expression in QgsDataDefined multiplies the states of of the class and this is not good IMHO. Same thing for the mField. I'd rather have only a QgsExpression (smart) pointer and a mActive flag, and use expression introspection (like QgsExpression::isSimpleField) instead of duplicating those functionalities in QgsDataDefined.

**Comment by vmora** [comment]
PS: I understand that QgsExpression is not copyable (I'm the one who initially disabled the buggy cpy), but implementing the cpy is IMO a cleaner option.

**Comment by nyalldawson** [comment]
@vmora it's on my TODO list (making QgsExpression & QgsDataDefined implicitly shared), but it's a long list ;)

**Comment by vmora** [comment]
@nyalldawson I'm asking because I'm modifying the code for QgsSymbolV2::sizeExpression and the rest, and the QgsDataDefined interface is tricky because an expressionString can be defined while expression() will return null nonetheless.

**Comment by vmora** [comment]
I'm trying to change the interface of setSizeExpression as you requested, but I'm stuck because I cannot return a pointer to a QgsDataDefined that does not exist in this case, and I certainly don't want to return an allocated value.

@nyalldawson can you tell me why you choose to pass QgsDataDefined by pointer with ownership transfer instead of passing by const & and copying.

**Comment by nyalldawson** [comment]
manually merged in 1e46196937827d2a79f978b0ed67e97be9531c81 - thanks @vmora!

**Comment by vmora** [comment]
The feature was funded by [Agence de l'Eau Adour Garonne](http://www.eau-adour-garonne.fr/fr/index.html).
